api/controllers/shabads.js

```diff
@@ -438,64 +437,91 @@ const getShabad = (ShabadIDQ, sinceDate = null) =>
         if (ShabadIDQLength > 1) {
-          multipleShabad = `FIELD(ShabadID, ${tokens}),`;
+          multipleShabadOrder = `FIELD(ShabadID, ${tokens}),`;
           parameters.push(...ShabadIDQ);
         }
 
         const q = `SELECT ${allColumns} ${allFrom}
                     WHERE s.ShabadID IN (${tokens}) ${allColumnsWhere} ${sinceQuery}
-                    ORDER BY ${multipleShabad} v.ID ASC`;
+                    ORDER BY ${multipleShabadOrder} v.ID ASC`;
 
         conn
           .query(q, parameters)
-          .then(rows => {
-            if (rows.length > 0) {
-              const shabadInfo = {
-                shabadId: rows[0].ShabadID,
-                pageNo: rows[0].PageNo,
-                source: lib.getSource(rows[0]),
-                raag: lib.getRaag(rows[0]),
-                writer: lib.getWriter(rows[0]),
+          .then(async rows => {
+            if (rows.length > 0 && ShabadIDQLength === 1) {
+              // single shabad
+              resolve(getShabadSingle(rows));
+            } else if (rows.length > 0) {
+              // multiple shabads
+              const output = {
+                shabadIds: [],
+                shabads: [],
               };
-
-              const verses = rows.map(lib.prepVerse);
-              const q1 = `(SELECT 'previous' as navigation,ShabadID FROM Shabad WHERE VerseID = ? LIMIT 1)
-              UNION
-            (SELECT 'next' as navigation,ShabadID FROM Shabad WHERE VerseID= ? LIMIT 1);`;
-              conn
-                .query(q1, [rows[0].ID - 1, rows[rows.length - 1].ID + 1])
-                .then(rows1 => {
-                  let previous = null;
-                  let next = null;
-                  rows1.forEach(row => {
-                    if (row.navigation === 'previous') {
-                      previous = row.ShabadID;
-                    }
-                    if (row.navigation === 'next') {
-                      next = row.ShabadID;
-                    }
-                  });
-                  const navigation = {
-                    previous,
-                    next,
-                  };
-
-                  resolve({
-                    shabadInfo,
-                    count: verses.length,
-                    navigation,
-                    verses,
-                  });
-                  conn.end();
-                })
-                .catch(err => reject(err));
+              let curShabadID = -1;
+              let counter = 0;
+              rows.forEach(row => {
+                if (row.ShabadID !== curShabadID) {
+                  curShabadID = row.ShabadID;
+                  output.shabadIds.push(row.ShabadID);
+                  counter = output.shabads.push([]) - 1;
+                }
+                output.shabads[counter].push(row);
+              });
+
+              const outputShabadPromises = output.shabads.map(getShabadSingle);
+              output.shabads = await Promise.all(outputShabadPromises);
```

Comment:
is this needed? `getShabadSingle` doesn't actually return a promise.

Reply:
I think it's needed, because of the async call to `getShabadNavigation`